KnapsackGA.cpp

```diff
@@ -0,0 +1,159 @@
+#include <vector>
+#include <iostream>
+#include <random>
+#include <algorithm>
+
+#include "GeneticAlgorithm.hpp"
+
+struct Solution
+{
+    std::vector<bool> itemSelection;
+    double fitness;
+};
+
+class KnapsackGA : public GeneticAlgorithm
+{
+public:
+    KnapsackGA(int populationSize, int generations, int elitSize, double mutationRate, double crossoverRate, const std::vector<int> &values, const std::vector<int> &weights, int knapsackCapacity)
+        : GeneticAlgorithm(populationSize, generations, elitSize, mutationRate, crossoverRate), values(values), weights(weights), knapsackCapacity(knapsackCapacity), distr(0, 1), distrInt(0, values.size() - 1)
+    {
+        solutions.resize(populationSize);
+    }
+
+    Solution getBestSolution() const
+    {
+        return solutions[0];
+    }
+
+private:
+    void initializePopulation() override
+    {
+        for (int i = 0; i < solutions.size(); ++i)
+        {
+            solutions[i] = generateSolution();
+        }
+    }
+
+    void calculateFitness() override
+    {
+        for (Solution &solution : solutions)
+        {
+            solution.fitness = fitness(solution);
+        }
+    }
+
+    void selection() override
+    {
+        sort();
+        solutions.resize(elitSize);
+    }
+
+    void crossover() override
+    {
+        for (size_t i = 0; i < elitSize; i += 2)
+        {
+            if (distr(gen) < crossoverRate)
+            {
+                int crossoverPoint = distrInt(gen);
+
+                Solution child1;
+                Solution child2;
+
+                child1.itemSelection.insert(child1.itemSelection.end(), solutions[i].itemSelection.begin(), solutions[i].itemSelection.begin() + crossoverPoint);
+                child1.itemSelection.insert(child1.itemSelection.end(), solutions[i + 1].itemSelection.begin() + crossoverPoint, solutions[i + 1].itemSelection.end());
+
+                child2.itemSelection.insert(child2.itemSelection.end(), solutions[i + 1].itemSelection.begin(), solutions[i + 1].itemSelection.begin() + crossoverPoint);
+                child2.itemSelection.insert(child2.itemSelection.end(), solutions[i].itemSelection.begin() + crossoverPoint, solutions[i].itemSelection.end());
+
+                solutions.push_back(child1);
+                solutions.push_back(child2);
+            }
+        }
+    }
+
+    void mutation() override
+    {
+        for (size_t i = elitSize; i < solutions.size(); ++i)
+        {
+            if (distr(gen) < mutationRate)
+            {
+                int mutationPoint = distrInt(gen);
+                solutions[i].itemSelection[mutationPoint] = !solutions[i].itemSelection[mutationPoint];
+            }
+        }
+    }
```

Comment:
Ez is ugyanazon a generacion dolgozik. Tehat lehet, hogy a mutaltat fogja tovabb mutalni, vagy crossoverelni, vagy akarmi.